util/src/main/scala/Money.scala

```diff
@@ -128,25 +202,248 @@ object Money {
+    fromDecimalAmount(amount, Currency.getInstance(currency))(BigDecimal.RoundingMode.HALF_EVEN)
+
+  def EUR(amount: BigDecimal): Money = decimalAmountWithCurrencyAndHalfEvenRounding(amount, "EUR")
+  def USD(amount: BigDecimal): Money = decimalAmountWithCurrencyAndHalfEvenRounding(amount, "USD")
+  def GBP(amount: BigDecimal): Money = decimalAmountWithCurrencyAndHalfEvenRounding(amount, "GBP")
+  def JPY(amount: BigDecimal): Money = decimalAmountWithCurrencyAndHalfEvenRounding(amount, "JPY")
+
+  val TypeName = "centAmount"
 
-  /** Creates a new Money object, forcing the scale according to the currency. */
-  def make(amount: BigDecimal, currency: Currency, mode: RoundingMode): Money =
+  def fromDecimalAmount(amount: BigDecimal, currency: Currency)(implicit mode: RoundingMode) =
     Money(amount.setScale(currency.getDefaultFractionDigits, mode), currency)
 
-  def make(amount: BigDecimal, currency: Currency): Money =
-    make(amount, currency, BigDecimal.RoundingMode.HALF_EVEN)
+  def fromCentAmount(centAmount: Long, currency: Currency) = {
+    val centFactor = BigDecimal(1) / BigDecimal(10).pow(currency.getDefaultFractionDigits)
+    val amount = BigDecimal(centAmount) * centFactor
 
-  implicit def moneyMonoid(implicit c: Currency): Monoid[Money] = new Monoid[Money] {
-    def combine(x: Money, y: Money): Money = x + y
-    val empty: Money = Money.make(0, c)
+    fromDecimalAmount(amount, currency)(BigDecimal.RoundingMode.UNNECESSARY)
   }
 
+  def zero(currency: Currency) =
+    fromCentAmount(0L, currency)
+
   implicit def moneyMonoid(implicit c: Currency, mode: RoundingMode): Monoid[Money] = new Monoid[Money] {
     def combine(x: Money, y: Money): Money = x + y
-    val empty: Money = Money.make(0, c, mode)
+    val empty: Money = Money.zero(c)
+  }
+}
+
+case class HighPrecisionMoney(amount: BigDecimal, fractionDigits: Int, centAmount: Long, currency: Currency) extends BaseMoney with Ordered[Money] {
+  import HighPrecisionMoney._
+
+  require(amount.scale == fractionDigits,
+    "The scale of the given amount does not match the scale of the provided currency." +
+    " - " + amount.scale + " <-> " + fractionDigits)
+
+  require(fractionDigits >= currency.getDefaultFractionDigits,
+    "`fractionDigits` should be  >= than the default fraction digits of the currency.")
+
+  val `type` = TypeName
+
+  lazy val preciseAmountAsLong: Long =
+    (amount * BigDecimal(10).pow(fractionDigits)).toLongExact
+
+  def withFractionDigits(fd: Int)(implicit mode: RoundingMode) = {
+    val newAmount = amount.setScale(fd, mode)
+
+    HighPrecisionMoney(newAmount, fd, roundToCents(newAmount, currency), currency)
+  }
+
+  def + (other: HighPrecisionMoney)(implicit mode: RoundingMode): HighPrecisionMoney =
+    calc(this, other, _ + _)
+
+  def + (m: Money)(implicit mode: RoundingMode): HighPrecisionMoney =
+    this + m.toHighPrecisionMoney(fractionDigits)
+
+  def + (money: BaseMoney)(implicit mode: RoundingMode): BaseMoney = money match {
```

Comment:
Since both options return `HighPrecisionMoney`, this function could also return it (instead of `BaseMoney`). Same below.

Reply:
Done: https://github.com/sphereio/sphere-scala-libs/pull/34/commits/611e9a3d546ebe4f682bfe2ffd8c6cd7f390ca2f